### PR TITLE
Handle nested modules

### PR DESCRIPTION
**Pull request by rofinn**
Currently, using nested modules can result in an `UndefVarError`.

```
julia> using Mocking

julia> Mocking.enable(; force=true)
Warning: Using experimental code which modifies jl_options global struct

julia> module A

       module B

       abstract type AbstractFoo end

       struct Foo <: AbstractFoo
           x::String
       end

       end

       bar(f::B.AbstractFoo) = println("blah")

       end
A

julia> import A: bar

julia> patch = @patch bar(f::B.AbstractFoo) = println("blar")
Mocking.Patch(:(bar(f::A.B.AbstractFoo)), ##665, Set(Any[:(A.B)]))

julia> Mocking.apply(patch) do
           bar(B.Foo("Blar"))
       end
ERROR: UndefVarError: A not defined
Stacktrace:
 [1] Mocking.PatchEnv(::Mocking.Patch, ::Bool) at /Users/rory/.playground/share/mocking/packages/v0.6/Mocking/src/Mocking.jl:156
 [2] apply(::Function, ::Mocking.Patch) at /Users/rory/.playground/share/mocking/packages/v0.6/Mocking/src/Mocking.jl:184
```

This is because the patch is only importing the top level module while the method is using a fully specified type signature:

```
ulia> Mocking.convert(Expr, patch)
quote
    import A.B
    bar(f::A.B.AbstractFoo) = begin
            (##662)(f)
        end
end
```

The solution should be to just recursively import all the parent modules.

### CODE REVIEW

File: test/patch.jl [hunk 1]
```diff
@@ -13,6 +13,26 @@ function strip_lineno!(expr::Expr)
...
+    x::String
+end
+
+end
+
+bar(f::ModB.AbstractFoo) = println("blah")
+
+end
+
+import ModA
```
Comment (ararslan) [inline test/patch.jl]: For 0.7, this will need to be
```julia
if VERSION >= v"0.7.0-DEV.1877"
    import .ModA
    import .ModA: bar, ModB
else
    import ModA
    import ModA: bar, ModB
end
```
because of https://github.com/JuliaLang/julia/pull/23579

File: test/patch.jl [hunk 2]
```diff
@@ -13,6 +13,29 @@ function strip_lineno!(expr::Expr)
...
+
+module ModB
+
+abstract type AbstractFoo end
+
+struct Foo <: AbstractFoo
+    x::String
+end
+
+end
```
Comment (omus) [inline test/patch.jl]: Can we indent the submodule for clarity?

Comment (iamed2) [reply test/patch.jl]: Or even just `end # ModA` comments

File: src/Mocking.jl [hunk 3]
```diff
@@ -53,7 +53,12 @@ struct Patch
         modules = Set()
         for b in values(trans)
             if isa(b, Expr)
-                push!(modules, b.args[1])
+                m = b.args[1]
+                while isa(m, Expr)
+                    push!(modules, m)
+                    m = m.args[1]
+                end
+                push!(modules, m)
```
Comment (omus) [inline src/Mocking.jl]: Why the change?

Comment (omus) [reply src/Mocking.jl]: lol, you fixed this in a very different way than I thought you would. XD

File: test/patch.jl [hunk 4]
```diff
@@ -78,8 +101,17 @@ end
         ]
         for p in patches
             @test p.signature == :(f(h::Core.Int64=$RAND_EXPR(Core.Int64)))
-            @test p.modules == Set([:Core, RAND_MOD_EXPR])
+            @test p.modules == Set([:Core, :Base, RAND_MOD_EXPR])
```
Comment (omus) [inline test/patch.jl]: https://travis-ci.org/invenia/Mocking.jl/jobs/402418514#L509

**Comment by rofinn** [comment]
I guess we'll need to start special casing `Main` on 0.7 now?

**Comment by ararslan** [comment]
Could you not use `fullname` to get the full path to a module rather than having to use a relative reference? Also `import Main`, while a strange thing to do, does work in 0.7, at least at the REPL.

**Comment by iamed2** [comment]
> Also import Main, while a strange thing to do, does work in 0.7, at least at the REPL.

It works when you're in Main, like at the REPL. If you're in a package, it won't work, because each package has its own top level module (e.g., `parentmodule(Compat) == Compat`).
